chromadb/auth/fastapi.py

```diff
@@ -0,0 +1,121 @@
+
+from chromadb.config import System
+from chromadb.auth import (
+    ServerAuthenticationRequest,
+    AuthInfoType,
+    ServerAuthenticationResponse,
+    ServerAuthProvider,
+    ChromaAuthMiddleware,
+)
+from chromadb.auth.registry import resolve_provider
+
+logger = logging.getLogger(__name__)
+
+
+class FastAPIServerAuthenticationRequest(ServerAuthenticationRequest[Optional[str]]):
+    def __init__(self, request: Request) -> None:
+        self._request = request
+
+    @override
+    def get_auth_info(
+        self, auth_info_type: AuthInfoType, auth_info_id: Optional[str] = None
+    ) -> Optional[str]:
+        if auth_info_type == AuthInfoType.HEADER:
+            return str(self._request.headers[auth_info_id])
+        elif auth_info_type == AuthInfoType.COOKIE:
+            return str(self._request.cookies[auth_info_id])
+        elif auth_info_type == AuthInfoType.URL:
+            return str(self._request.query_params[auth_info_id])
+        elif auth_info_type == AuthInfoType.METADATA:
+            raise ValueError("Metadata not supported for FastAPI")
+        else:
+            raise ValueError(f"Unknown auth info type: {auth_info_type}")
+
+
+class FastAPIServerAuthenticationResponse(ServerAuthenticationResponse):
+    _auth_success: bool
+
+    def __init__(self, auth_success: bool) -> None:
+        self._auth_success = auth_success
+
+    @override
+    def success(self) -> bool:
+        return self._auth_success
+
+
+class FastAPIChromaAuthMiddleware(ChromaAuthMiddleware):
+    _auth_provider: ServerAuthProvider
+
+    def __init__(self, system: System) -> None:
+        super().__init__(system)
+        self._system = system
+        self._settings = system.settings
+        self._settings.require("chroma_server_auth_provider")
+        self._ignore_auth_paths: Dict[
+            str, List[str]
+        ] = self._settings.chroma_server_auth_ignore_paths
+        if self._settings.chroma_server_auth_provider:
+            logger.debug(
+                f"Server Auth Provider: {self._settings.chroma_server_auth_provider}"
+            )
+            _cls = resolve_provider(
+                self._settings.chroma_server_auth_provider, ServerAuthProvider
+            )
+            self._auth_provider = cast(ServerAuthProvider, self.require(_cls))
+
+    @override
+    def authenticate(
+        self, request: ServerAuthenticationRequest[Any]
+    ) -> Optional[ServerAuthenticationResponse]:
+        return FastAPIServerAuthenticationResponse(
+            self._auth_provider.authenticate(request)
+        )
+
+    @override
+    def ignore_operation(self, verb: str, path: str) -> bool:
+        if (
+            path in self._ignore_auth_paths.keys()
+            and verb.upper() in self._ignore_auth_paths[path]
+        ):
+            logger.debug(f"Skipping auth for path {path} and method {verb}")
+            return True
+        return False
+
+    @override
+    def instrument_server(self, app: ASGIApp) -> None:
```

Comment:
unused abstraction - remove

Reply:
gonna use this in the authZ part (coming up next). Also, if we want to implement basic auth with a challenge-response style can also be useful.